### DISCUSSION TITLE
Demo AWS CloudWatch Agent for AWS Xray

### DISCUSSION THREAD

**Discussion by hungran** [vntechies/blog#40] [Ideas]
Since AWS announcement  CloudWatch Agent adds support for OpenTelemetry traces and AWS X-Ray  https://aws.amazon.com/vi/about-aws/whats-new/2023/08/amazon-cloudwatch-agent-opentelemetry-traces-x-ray/
as public community, we should have a public demo for that

**Comment by hungran** [comment]
https://github.com/vntechies/017-aws-cwagent-xray
idea
- [x] Xray SDK for DynamoDB
- [ ] using OpenTelemetry along with Xray SDK
- [ ] Docs
   - [ ] Diagram draw.io
   - [ ] Readme.md
- [ ] Add more demo function that can be show on Trace: Eg S3, SNS
- [ ] Terraform to quickly install 
   - [ ] Create DynamoDB (table name: `message`)
   - [ ] IAM Role (SSM Managed Instance Core, Xray Permission, S3 Permission)
   - [ ] Security group (allow 8080)
   - [ ] CloudWatch Agent installation by Userdata
   - [ ] CWAgent configuration file should be store in SSM ParameterStore

**Comment by hungran** [comment]
Sample Configuration CWAgent:
```json
{
    "agent": {
        "run_as_user": "cwagent"
    },
    "traces": {
        "traces_collected": {
            "xray": {
                "bind_address": "127.0.0.1:2000",
                "tcp_proxy": {
                    "bind_address": "127.0.0.1:2000"
                }
            },
            "otlp": {}
        }
    },
    "metrics": {
        "metrics_collected": {
            "mem": {
                "measurement": [
                    "mem_used_percent"
                ]
            },
            "disk": {
                "measurement": [
                    "used_percent"
                ],
                "resources": [
                    "*"
                ]
            }
        },
        "append_dimensions": {
            "ImageId": "${aws:ImageId}",
            "InstanceId": "${aws:InstanceId}",
            "InstanceType": "${aws:InstanceType}",
            "AutoScalingGroupName": "${aws:AutoScalingGroupName}"
        }
    }
}
```

**Comment by hungran** [comment]
IAM Instance Profile:
- [AWSXRayDaemonWriteAccess](https://us-east-1.console.aws.amazon.com/iamv2/home#/policies/details/arn%3Aaws%3Aiam%3A%3Aaws%3Apolicy%2FAWSXRayDaemonWriteAccess)
- [AmazonSSMManagedInstanceCore](https://us-east-1.console.aws.amazon.com/iamv2/home#/policies/details/arn%3Aaws%3Aiam%3A%3Aaws%3Apolicy%2FAmazonSSMManagedInstanceCore)

**Comment by dinhdangkhoa0201** [comment]
![image](https://github.com/vntechies/blog/assets/45093050/3c293722-3426-4a0f-999b-dd333c3a08a2)
Outline of demo

**Reply by hungran** [reply]
@dinhdangkhoa0201 thanks, can we break project task base on this?

**Comment by dinhdangkhoa0201** [comment]
This is my project
https://github.com/dinhdangkhoa0201/amazon-cloadwatch-for-aws-xray.git
@hungran

**Comment by qmauvnt** [comment]
@hungran @dinhdangkhoa0201 do we have any plan to follow-up on this one?
